File: admin_panel/lib/api/banners.ts
import { apiFetch, getApiConfig } from "@/lib/api/config";

export interface BannerItem {
  id: string;
  title: string;
  message: string;
  image_url: string;
  price_label: string;
  course_id: string | null;
  telegram: string;
  is_active: boolean;
}

function headers() {
  const { adminApiKey } = getApiConfig();
  return {
    "Content-Type": "application/json",
    ...(adminApiKey ? { "x-admin-api-key": adminApiKey } : {}),
  };
}

async function parseError(response: Response, fallback: string) {
  try {
    const data = (await response.json()) as { detail?: string };
    return data.detail?.trim() || fallback;
  } catch {
    return fallback;
  }
}

export async function fetchBanners() {
  const response = await apiFetch("/api/v1/banners", { cache: "no-store" });
  if (!response.ok) throw new Error(await parseError(response, "Reklamalarni olishda xatolik."));
  const data = (await response.json()) as { items: BannerItem[] };
  return data.items ?? [];
}

export async function createBanner(payload: {
  title: string;
  message: string;
  image_url: string;
  price_label: string;
  course_id?: string | null;
  telegram: string;
  is_active?: boolean;
}) {
  const response = await apiFetch("/api/v1/banners", {
    method: "POST",
    headers: headers(),
    body: JSON.stringify(payload),
  });
  if (!response.ok) throw new Error(await parseError(response, "Reklama qo'shilmadi."));
  return (await response.json()) as BannerItem;
}

export async function updateBanner(id: string, payload: Partial<Omit<BannerItem, "id">>) {
  const response = await apiFetch(`/api/v1/banners/${id}`, {
    method: "PATCH",
    headers: headers(),
    body: JSON.stringify(payload),
  });
  if (!response.ok) throw new Error(await parseError(response, "Reklama yangilanmadi."));
  return (await response.json()) as BannerItem;
}

export async function removeBanner(id: string) {
  const response = await apiFetch(`/api/v1/banners/${id}`, {
    method: "DELETE",
    headers: headers(),
  });
  if (!response.ok) throw new Error(await parseError(response, "Reklama o'chirilmadi."));
}

File: backend/fastapi/app/schemas/banners.py
from datetime import datetime

from pydantic import BaseModel, Field


class BannerCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    message: str = ""
    image_url: str = ""
    price_label: str = ""
    course_id: str | None = None
    telegram: str = "Neuroscienceadmin"
    is_active: bool = False


class BannerUpdate(BaseModel):
    title: str | None = None
    message: str | None = None
    image_url: str | None = None
    price_label: str | None = None
    course_id: str | None = None
    telegram: str | None = None
    is_active: bool | None = None


class BannerItem(BaseModel):
    id: str
    title: str
    message: str
    image_url: str
    price_label: str
    course_id: str | None
    telegram: str
    is_active: bool
    created_at: datetime


class BannerListResponse(BaseModel):
    items: list[BannerItem]
